### PR TITLE
#473 serveless implementation using aws compute engine and serverless fram…

### PR DESCRIPTION
**Pull request by dheeraj-mummareddy**
**serveless implementation using aws compute engine and serverless framework**

- serverless example implementation using aws compute engine and serverless framework
- function creates an info endpoint which provides data from lambda context
- resolves: #473 Serverless pattern

Pull request description

- Implementation and documentation using aws compute engine and serverless framework

### CODE REVIEW

**Comment by roman-marcu** [comment]
Hi, AWS is a Serverless FaaS technique, but what about BaaS? According to https://martinfowler.com/articles/serverless.html serverless is not just about moving server side code into some ephemeral function, but also integration of third party remote application services directly into the front-end of our app, as an example for BaaS https://firebase.google.com/docs/database/.

It would be nice to add a gateway API and to do 2 calls, one for AWS(FaaS) and other for firebase, or others,(BaaS).

**Comment by dheeraj-mummareddy** [comment]
@roman-marcu acknowledge, will add BaaS endpoint... API gateway is already configured in the stack which is created by serverless framework, will expand documentation, please take a look at http section in serverless yaml 
https://serverless.com/framework/docs/providers/aws/events/apigateway/

**Comment by iluwatar** [comment]
@dheeraj-mummareddy the pull request has conflicts. Can you check it?

**Comment by dheeraj-mummareddy** [comment]
@iluwatar the PR is ready for review, the build passes locally but is not healthy in travis and looks like its failing for all other commits too. Is someone looking at the failure?

[INFO] throttling ......................................... FAILURE [  2.287 s]

**Comment by iluwatar** [comment]
It is a known issue https://github.com/iluwatar/java-design-patterns/issues/643, still unresolved

File: serverless/README.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+---
+layout: pattern
+title: serverless
+folder: serverless
+permalink: /patterns/serverless/
+categories: Architectural
+tags:
+ - Java
+ - Difficulty-Intermittent
```
Comment (iluwatar) [inline serverless/README.md]: Difficulty-Intermediate?

File: serverless/README.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    * resilient routing
+    * concurrent execution
+* Coordinated calls
+    * Decoupled via APIs
+    * scale-matched downstream
+    * secured
+* Serviceful operations
+    * Automated operations
+    * monitored applications
+    * Innovation mindset
```
Comment (iluwatar) [inline serverless/README.md]: Any links that would explain more about these katas? The terms alone are not enough.

File: serverless/README.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+using aws as a compute engine is our first choice.
+
+## (Backend as a Service or "BaaS")
+This example creates a backend for ‘persons’ collection which uses DynamoDB NoSQL 
+database service also provided by Amazon. 
+
+## AWS lambda function implementation
+
+AWS lambda SDK provides pre-defined interface `com.amazonaws.services.lambda.runtime
+.RequestHandler` to implement our lambda function.
```
Comment (iluwatar) [inline serverless/README.md]: Link to AWS lambda SDK?

File: serverless/README.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+* `serverless deploy --stage=dev --verbose`
+
+Based on the configuration in `serverless.yml` serverless framework creates following 
+resources
+* cloud formation stack for S3 (ServerlessDeploymentBucket)
+* IAM Role (IamRoleLambdaExecution)
+* cloud watch (log groups)
+* API Gateway (ApiGatewayRestApi) 
+* Lambda function
+* DynamoDB collection
```
Comment (iluwatar) [inline serverless/README.md]: Spelling, cloud formation -> CloudFormation, cloud watch -> CloudWatch

File: serverless/src/main/java/com/iluwatar/serverless/baas/model/Address.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
+package com.iluwatar.serverless.baas.model;
+
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBAttribute;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBDocument;
+
+import java.io.Serializable;
+
+/**
+ * Address Object
```
Comment (iluwatar) [inline serverless/src/main/java/com/iluwatar/serverless/baas/model/Address.java]: Address class, not an object

File: serverless/src/main/java/com/iluwatar/serverless/baas/model/Person.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBAttribute;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBAutoGeneratedKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBHashKey;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBTable;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.io.Serializable;
+
+/**
+ * Person Request
```
Comment (iluwatar) [inline serverless/src/main/java/com/iluwatar/serverless/baas/model/Person.java]: Person class, not a request

File: serverless/src/test/java/com/illuwatar/serverless/baas/api/FindPersonApiHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+
+import java.util.Collections;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+/**
+ * Created by dheeraj.mummar on 3/5/18.
```
Comment (iluwatar) [inline serverless/src/test/java/com/illuwatar/serverless/baas/api/FindPersonApiHandlerTest.java]: Explain what the class does

File: serverless/src/test/java/com/illuwatar/serverless/baas/api/SavePersonApiHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.mockito.Mockito.*;
+
+/**
+ * Created by dheeraj.mummar on 3/4/18.
```
Comment (iluwatar) [inline serverless/src/test/java/com/illuwatar/serverless/baas/api/SavePersonApiHandlerTest.java]: Explain what the class does

File: serverless/src/test/java/com/iluwatar/serverless/faas/api/LambdaInfoApiHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsNull.notNullValue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * LambdaInfoApiHandlerTest
```
Comment (iluwatar) [inline serverless/src/test/java/com/iluwatar/serverless/faas/api/LambdaInfoApiHandlerTest.java]: Explain what the class does

File: serverless/README.md (hunk 1 above)
Comment (dheeraj-mummareddy) [reply serverless/README.md]: @iluwatar I feel like the difficulty level is Intermediate based on the complexity to adopt, but let me know if your thoughts?

**Comment by iluwatar** [comment]
@dheeraj-mummareddy are you working on this? Please comment when you want another review.

File: serverless/README.md (hunk 1 above)
Comment (iluwatar) [reply serverless/README.md]: I mean literally, Difficulty-Intermittent is wrong...

File: serverless/pom.xml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>serverless</artifactId>
+    <parent>
+        <groupId>com.iluwatar</groupId>
+        <artifactId>java-design-patterns</artifactId>
+        <version>1.19.0-SNAPSHOT</version>
```
Comment (iluwatar) [inline serverless/pom.xml]: Should be updated to `1.20.0-SNAPSHOT`

**Comment by iluwatar** [comment]
@dheeraj-mummareddy please comment when the (minor) changes are done so we can proceed with the merge

File: serverless/README.md (hunk 1 above)
Comment (dheeraj-mummareddy) [reply serverless/README.md]: @iluwatar good catch, I think I copied the initial template from EIP patter, yup...will fix that with an other PR.

**Comment by dheeraj-mummareddy** [comment]
@iluwatar change have been fixed.

**Comment by iluwatar** [comment]
Good work @dheeraj-mummareddy thank you very much for this contribution :+1:

**Comment by iluwatar** [comment]
@all-contributors please add @dheeraj-mummareddy for code

**Pull request merged**